Sort mempool by min(feerate, ancestor_feerate)

sdaftuar opened the pull request:
This more closely approximates the desirability of a given transaction for
mining, and should result in less re-sorting when transactions get removed from
the mempool after being mined.

I measured this as approximately a 5% speedup in removeForBlock.

promag commented on src/txmempool.h:
```diff
@@ -261,27 +261,44 @@ class CompareTxMemPoolEntryByEntryTime
+    // Picks min(feerate, feerate_with_ancestors)
+    bool UseAncestorScore(const CTxMemPoolEntry &a) const
+    {
+        double f1 = (double)a.GetModifiedFee() * a.GetSizeWithAncestors();
```
Remove double cast?

promag commented on src/txmempool.h:
```diff
@@ -261,27 +261,44 @@ class CompareTxMemPoolEntryByEntryTime
+        bool fUseAAncestors = UseAncestorScore(a);
+        bool fUseBAncestors = UseAncestorScore(b);
+
+        double aModFee = fUseAAncestors ? a.GetModFeesWithAncestors() : a.GetModifiedFee();
```
Why floating point? Aren't these all 64 integers?

promag commented on src/txmempool.h:
```diff
@@ -261,27 +261,44 @@ class CompareTxMemPoolEntryByEntryTime
 
-        double bFees = b.GetModFeesWithAncestors();
-        double bSize = b.GetSizeWithAncestors();
+        double bModFee = fUseBAncestors ? b.GetModFeesWithAncestors() : b.GetModifiedFee();
```
Instead of factoring out `UseAncestorScore`, do (missing comments):
```cpp
void GetModFeeAndSize(const CTxMemPoolEntry& a, int64_t& mod_fee, int64_t& size) {
    int64_t f1 = a.GetModifiedFee() * a.GetSizeWithAncestors();
    int64_t f2 = a.GetModFeesWithAncestors() * a.GetTxSize();
    if (f1 > f2) {
        mod_fee = a.GetModFeesWithAncestors();
        size = a.GetSizeWithAncestors();
    } else {
        mod_fee = a.GetModifiedFee();
        size = a.GetTxSize();
    }
}
```

promag commented on src/txmempool.h:
```diff
@@ -261,27 +261,44 @@ class CompareTxMemPoolEntryByEntryTime
     }
+
+    // Calculate which score to use for an entry (avoiding division).
+    // Picks min(feerate, feerate_with_ancestors)
```
Nit, missing `.`.

sdaftuar commented:
@promag Thanks for taking a look.  I mimicked `CompareTxMempoolEntryByDescendantScore` in this PR to make review easier, since the two concepts have a symmetry.

The use of `double` for the intermediate value is to prevent overflow of an int64_t, which is theoretically possible.  That's also why there's an explicit double cast before multiplication.

I'll rewrite the helper function along the lines of your suggestion.

sdaftuar committed:
Refactor CompareTxMemPoolEntryByDescendantScore

sdaftuar committed:
Sort mempool by min(feerate, ancestor_feerate)

This more closely approximates the desirability of a given transaction for
mining.

promag commented:
> The use of double for the intermediate value is to prevent overflow of an int64_t, which is theoretically possible.

How about http://releases.llvm.org/4.0.0/tools/clang/docs/LanguageExtensions.html#checked-arithmetic-builtins?

promag commented on src/txmempool.h:
```diff
@@ -206,31 +206,36 @@ class CompareTxMemPoolEntryByDescendantScore
-    // Calculate which score to use for an entry (avoiding division).
-    bool UseDescendantScore(const CTxMemPoolEntry &a) const
+    // Return the fee/size we're using for sorting this entry.
+    void GetModFeeAndSize(const CTxMemPoolEntry &a, double &mod_fee, double &size) const
```
& next to type.

promag commented on src/txmempool.h:
```diff
@@ -261,27 +266,46 @@ class CompareTxMemPoolEntryByEntryTime
     }
+
+    // Return the fee/size we're using for sorting this entry.
+    void GetModFeeAndSize(const CTxMemPoolEntry &a, double &mod_fee, double &size) const
```
Same.

sdaftuar commented:
@promag The style of where to put the `&` doesn't seem to be in our style guide and is inconsistent already in this file, so I'm going to leave it the way I have it (which happens to be my preference for where to put the `&` :) ).

I don't think it's a problem for code-correctness if the arithmetic overflows, so I'd rather just handle it as we currently do than introduce complication from calculating things two ways.

promag commented:
utACK 9a51319.

> I'm going to leave it the way I have it

Fine.

> so I'd rather just handle it as we currently do

It was just curiosity. Maybe `GetModFeeAndSize` values could be cached in `CTxMemPoolEntry` to improve even further? (adding 128 bytes per entry)

ryanofsky reviewed (commented):
utACK 9a51319578091234fdd218a1eb144d517ea82b85

> I measured this as approximately a 5% speedup in removeForBlock.

I'm always interested to know how you benchmark these things, in case you want to post details.

> The style of where to put the & doesn't seem to be in our style guide

Not important, but we do currently use the opposite setting for the code formatter:

https://github.com/bitcoin/bitcoin/blob/45173fa6fca9537abb0a0554f731d14b9f89c456/src/.clang-format#L40

sdaftuar commented:
>>I measured this as approximately a 5% speedup in removeForBlock.

>I'm always interested to know how you benchmark these things, in case you want to post details.

@ryanofsky I ran through nearly 3 months of historical block & transaction data (roughly March-May 2017), which I record and can play back to a patched bitcoind.  I ran two historical studies, one with this patch and one without, each with added benchmarks to specifically measure the time spent in removeForBlock.

TheBlueMatt commented:
utACK 9a51319578091234fdd218a1eb144d517ea82b85

gmaxwell commented:
ACK

jimpo commented:
Maybe augment the `MempoolAncestorIndexingTest` with a case of a higher fee parent with a lower fee child (for which the sort order has changed).

sdaftuar committed:
Add test for new ancestor feerate sort behavior

sdaftuar committed:
Use mempool's ancestor sort in transaction selection

Transaction selection for mining tracks ancestor feerates that are
modified based on transactions that have already been selected.  This
commit de-duplicates the code so that the ancestor feerate sorting used
by the mempool can also be directly applied to the miner.

sdaftuar commented:
I just remembered that the mining code had this ancestor feerate sorting logic duplicated; I just added a commit that attempts to re-use the mempool code so that the new behavior will apply there as well.

TheBlueMatt reviewed (commented):
uACK 0a22a52918ad5af6d105b4f5ae9dd6c52199f0e8

laanwj commented:
utACK 0a22a5291

promag commented:
re-utACK 0a22a52. Added test looks good.

sdaftuar commented on src/miner.cpp:
```diff
@@ -352,7 +352,7 @@ void BlockAssembler::addPackageTxs(int &nPackagesSelected, int &nDescendantsUpda
             iter = mempool.mapTx.project<0>(mi);
             if (modit != mapModifiedTx.get<ancestor_score>().end() &&
-                    CompareModifiedEntry()(*modit, CTxMemPoolModifiedEntry(iter))) {
+                    CompareTxMemPoolEntryByAncestorFee()(*modit, CTxMemPoolModifiedEntry(iter))) {
```
Yes -- this change was only because I wanted to delete the code for CompareModifiedEntry, which is otherwise no longer needed.
